pkg/sys/socket_unix.go

```
@@ -49,6 +49,11 @@ func GetLocalListener(path string, uid, gid int) (net.Listener, error) {
 		return nil, err
 	}
 
+	fileInfo, err := os.Stat(path)
+	if err == nil && fileInfo.IsDir() {
+		os.RemoveAll(path)
```

This sounds scary. Probably better to just return an error.

---

done, thanks. @AkihiroSuda

---

I meant to just return an error without removing the dir

---

The directory was accidentally created by containerd, I think we should delete it and make containerd can start successfully. @AkihiroSuda

---

or we should prevent to create dir /run/containerd/containerd.sock  in WithMounts .It looks ugly,
```
func WithMounts(osi osinterface.OS, config *runtime.ContainerConfig, extra []*runtime.Mount, mountLabel string) oci.SpecOpts {
		for _, mount := range mounts {
			var (
				dst = mount.GetContainerPath()
				src = mount.GetHostPath()
			)
			// Create the host path if it doesn't exist.
			// TODO(random-liu): Add CRI validation test for this case.
			if _, err := osi.Stat(src); err != nil {
				if !os.IsNotExist(err) {
					return fmt.Errorf("failed to stat %q: %w", src, err)
				}
				if err := osi.MkdirAll(src, 0755); err != nil {
					return fmt.Errorf("failed to mkdir %q: %w", src, err)
				}
			}
```

---

interesting I find containerd.sock.ttrpc will not deleted.
grpc use 
```
func (l *listenSocket) Close() error {
	err := l.Listener.Close()
	channelz.RemoveEntry(l.channelzID)
	channelz.Info(logger, l.channelzID, "ListenSocket deleted")
	return err
}
```
call 
```
func (ln *UnixListener) close() error {
	// The operating system doesn't clean up
	// the file that announcing created, so
	// we have to clean it up ourselves.
	// There's a race here--we can't know for
	// sure whether someone else has come along
	// and replaced our socket name already--
	// but this sequence (remove then close)
	// is at least compatible with the auto-remove
	// sequence in ListenUnix. It's only non-Go
	// programs that can mess us up.
	// Even if there are racy calls to Close, we want to unlink only for the first one.
	ln.unlinkOnce.Do(func() {
		if ln.path[0] != '@' && ln.unlink {
			syscall.Unlink(ln.path)
		}
	})
	return ln.fd.Close()
}
```
will delete socket.